database2/ffldb/blockio.go

```diff
@@ -0,0 +1,749 @@
+	// write locks should only be held for the minimum time necessary.
+	obfMutex         sync.RWMutex
+	lruMutex         sync.Mutex
+	openBlocksLRU    *list.List // Contains uint32 block file numbers.
+	fileNumToLRUElem map[uint32]*list.Element
+	openBlockFiles   map[uint32]*lockableFile
+
+	// writeCursor houses the state for the current file and location that
+	// new blocks are written to.
+	writeCursor *writeCursor
+
+	// These functions are set to openFile, openWriteFile, and deleteFile by
+	// default, but are exposed here to allow the whitebox tests to replace
+	// them when working with mock files.
+	openFileFunc      func(fileNum uint32) (*lockableFile, error)
+	openWriteFileFunc func(fileNum uint32) (filer, error)
+	deleteFileFunc    func(fileNum uint32) error
+}
+
+// blockLocation identifies a particular block file and location.
+type blockLocation struct {
+	blockFileNum uint32
+	fileOffset   uint32
+	blockLen     uint32
+}
+
+// deserializeBlockLoc deserializes the passed serialized block location
+// information.  This is data stored into the block index metadata for each
+// block.  The serialized data passed to this function MUST be at least
+// blockLocSize bytes or it will panic.  The error check is avoided here because
+// this information will always be coming from the block index which includes a
+// checksum to detect corruption.  Thus it is safe to use this unchecked here.
+func deserializeBlockLoc(serializedLoc []byte) blockLocation {
+	// The serialized block location format is:
+	//
+	//  [0:4]  Block file (4 bytes)
+	//  [4:8]  File offset (4 bytes)
+	//  [8:12] Block length (4 bytes)
+	return blockLocation{
+		blockFileNum: byteOrder.Uint32(serializedLoc[0:4]),
+		fileOffset:   byteOrder.Uint32(serializedLoc[4:8]),
+		blockLen:     byteOrder.Uint32(serializedLoc[8:12]),
+	}
+}
+
+// serializeBlockLoc returns the serialization of the passed block location.
+// This is data to be stored into the block index metadata for each block.
+func serializeBlockLoc(loc blockLocation) []byte {
+	// The serialized block location format is:
+	//
+	//  [0:4]  Block file (4 bytes)
+	//  [4:8]  File offset (4 bytes)
+	//  [8:12] Block length (4 bytes)
+	var serializedData [12]byte
+	byteOrder.PutUint32(serializedData[0:4], loc.blockFileNum)
+	byteOrder.PutUint32(serializedData[4:8], loc.fileOffset)
+	byteOrder.PutUint32(serializedData[8:12], loc.blockLen)
+	return serializedData[:]
+}
+
+// blockFilePath return the file path for the provided block file number.
+func blockFilePath(dbPath string, fileNum uint32) string {
+	fileName := fmt.Sprintf(blockFilenameTemplate, fileNum)
+	return filepath.Join(dbPath, fileName)
+}
+
+// openWriteFile returns a file handle for the passed flat file number in
+// read/write mode.  The file will be created if needed.  It is typically used
+// for the current file that will have all new data appended.  Unlike openFile,
+// this function does not keep track the open file and it is not subject to the
```

Comment:
Typo: `does not keep track the open file`
Correction: `does not keep track *of* the open file`